python/Ice/async/client.py

```diff
@@ -3,55 +3,82 @@
 # Copyright (c) ZeroC, Inc. All rights reserved.
 #
 
+import asyncio
 import sys
-import traceback
 import Ice
 
 Ice.loadSlice('Hello.ice')
 import Demo
 
 
-class Callback:
-    def response(self, f):
-        try:
-            f.result()
-        except Demo.RequestCanceledException:
-            print("Demo.RequestCanceledException")
-        except Exception:
-            print("sayHello AMI call failed:")
-            traceback.print_exc()
-
+async def main():
 
-def run(communicator):
-    hello = Demo.HelloPrx.checkedCast(communicator.propertyToProxy('Hello.Proxy'))
-    if not hello:
-        print(sys.args[0] + ": invalid proxy")
-        sys.exit(1)
+    # Ice.initialize returns an initialized Ice communicator; the communicator is destroyed once it goes out of scope.
+    with Ice.initialize(sys.argv, "config.client") as communicator:
 
-    menu()
+        # The communicator initialization removes all Ice-related arguments from sys.argv
+        if len(sys.argv) > 1:
+            print(sys.argv[0] + ": too many arguments")
+            return 1
 
-    c = None
-    while c != 'x':
         try:
+            hello = Demo.HelloPrx.checkedCast(communicator.propertyToProxy('Hello.Proxy'))
+        except Ice.Exception as ex:
+            print("invalid proxy:", ex)
+            return 1
+
+        menu()
+
+        tasks = set()
+        done = set()
+
+        while True:
             sys.stdout.write("==> ")
             sys.stdout.flush()
-            c = sys.stdin.readline().strip()
-            if c == 'i':
-                hello.sayHello(0)
-            elif c == 'd':
-                cb = Callback()
-                hello.sayHelloAsync(5000).add_done_callback(cb.response)
-            elif c == 's':
-                hello.shutdown()
-            elif c == 'x':
-                pass  # Nothing to do
-            elif c == '?':
-                menu()
-            else:
-                print("unknown command `" + c + "'")
-                menu()
-        except Ice.Exception as ex:
-            print(ex)
+            nextCharTask = asyncio.create_task(getNextChar())
+            tasks.add(nextCharTask)
+            while nextCharTask not in done:
+                done, tasks = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
```

Comment:
It's not clear to me why we maintain a tasks set here instead of just directly awaiting on `getNextChar` (or just directly call await on `loop.run_in_executor(None, sys.stdin.readline)`) or await on the `sayHelloAsync` calls.

Reply:
I don't understand your comment: we simply launch a number of tasks and await them, i.e. let them run "in parallel".